Cargo profile dependencies

Manishearth opened the pull request:
Tracking issue: https://github.com/rust-lang/rust/issues/48683

[Rendered](https://github.com/Manishearth/rfcs/blob/custom-profiles/text/2282-profile-dependencies.md)

Manishearth commented:
pre-rfc at https://internals.rust-lang.org/t/strawman-pre-rfc-custom-profiles/6412/9

matklad commented:
>Custom profiles can be listed in a .cargo/config, however the user is responsible for clearing up build directories if the profile changes. That is, it is undefined behavior to run cargo build --profile foo if foo has been defined in .cargo/config and the profile has been edited since the last time you ran cargo build --profile foo.

Hm, I would think Cargo should just rebuild everything in this case? And I even think this will happen naturally with the current implementation: IIRC, compiler flags are hashed into the fingerprint for determining if the crate should be rebuild. 

What is the purpose of profiles `.cargo/config`? Is it something for the "big build system" use case? One interesting idea I have is that we can allow to specify profile as json on the command line. That would allow the build system to control Cargo without writing any files.

Manishearth commented:
> Hm, I would think Cargo should just rebuild everything in this case? 

I think tracking changes to config files might be out of scope for cargo

> And I even think this will happen naturally with the current implementation: IIRC, compiler flags are hashed into the fingerprint for determining if the crate should be rebuild.

If this is true then that works great!

> What is the purpose of profiles .cargo/config? Is it something for the "big build system" use case? 

Well, two purposes. Lets people specify global useful profiles like `cargo build --debugopt` or something.

Also, it fixes the big build system case -- Having a custom profile for each combination of flags quickly leads to blowup. Firefox lets you individually flip debug symbols, debug-assertions, and optimization levels. There are more involved, though. (There are use cases; folks hacking on Rust code may want different settings from folks hacking on C++ code)

Profiles aren't part of the package _anyway_, they're build info, and that really should go in `config` anyway.

est31 commented:
How is this compatible with the [profiles 2.0 proposal](https://github.com/rust-lang/cargo/issues/4140#issuecomment-306955783) by @matklad ? IIRC there was an embargo for new features for profiles until they get revamped? Generally I like the idea though.

ehuss commented on text/0000-custom-cargo-profiles.md:
```diff
@@ -0,0 +1,143 @@
+It is not possible to have the same crate compiled in different modes as a build dependency and a regular dependency within the same profile.
+
+
+`build_override` is itself 
```
Is this an unfinished statement?

Manishearth replied:
My bad, that line should be removed.

Manishearth commented:
@est31 I think this addresses the "some kind of DSL to map crates to flags".

This RFC does not attempt to handle the issue with test/bench/doc but I think that can be addressed independently. One solution for that would be closer to the "includeable profiles" proposal in the pre-RFC, and have something like `[profile.dev] testing = otherprofilename`

matklad commented:
@est31 I think of this RFC as basically profiles 2.0, and I really love it :) It doesn't solve the "which build in profile is run when" problem, but I think it can be addressed separately. Although I would prefer that instead of  `[profile.dev] testing = otherprofilename` we'd just killed all build-in profiles other then `dev` or `release`.

est31 commented on text/0000-custom-cargo-profiles.md:
```diff
@@ -0,0 +1,141 @@
+
+We amend this to add that you can define custom profiles with the `profile.foo` key syntax. These can be invoked via
+`cargo build --profile foo`. The `dev`/`doc`/`bench`/etc profiles remain special. Each custom profile, aside from the
+"special" ones, gets a folder in `target/`, named after the profile. "dev" and "debug" are considered to be aliases
```
Naming folders after profiles is no good idea because of profiles named `del` or similar.

matklad replied:
Profiles affect only the leaf crate, so I think it's ok not to guard against bad names here. That is, even if one uploads crate with a forbidden profile name to crates.io, no bad things should happen.

Although, it might be a good idea to give a warning/error like "Hey, you named your profile NUL, and this will probably break your crate on windows".

Manishearth replied:
yeah, I agree with @matklad here. We can add warnings for badly-named profiles (doesn't need to be in the rfc)

ben0x539 commented:
Hash the profile data into the output path and you don't need to worry about undefined behavior on profile changes, I guess.

jonhoo commented:
One question I have with this is how this applies transitively. For example, consider crate `c` which depends on `b` which depends on `a`. Say `b` knows that `a` should *always* be compiled with optimizations. If `c` is compiled in debug mode, is it necessary for `c` to *also* declare that `b.a` should be compiled with optimizations? Or can `c` somehow be informed by `b` that `a` should be compiled with optimizations?

matklad commented:
@jonhoo currently, only the leaf crate's profiles are accounted for (it can even be argued on this basis that profiles should not live in Cargo.toml and should go into `.cargo/config` instead), and, for workspaces, you can only specify profiles in the root of the workspace. I believe that this RFC does not propose to change this mechanics.

I can imagine an orthogonal RFC though, which allows upstream crate to *advice* some profile settings to the downstream crates. For example, the image crate could advise "compile me with -O 3 in the dev profile", and this setting would be used for crates which depend on image, unless they explicitly override it. 

The possible interaction between this RFC and hypothetical advise RFC is that it's not clear how upstream crates can advice on custom profiles (they don't even know their names!). I think it's reasonable to assume though that this is not a super important problem which we need to solve know: either of "advises do not work with custom profiles" or "custom profile can opt into advises from a particular build-in profile" seems like an ok solution.

matklad commented:
Also, cc @wycats for the whole discussion :)

jonhoo commented:
@matklad you're right that this could totally be addressed in a separate RFC. In fact, I wonder whether, if we had that RFC, that would alleviate many of the pains that this RFC tries to counter. The reason I bring this up as a concern here is that I worry that we'll see these extra profiles added all over the place as a result of this particular solution to the problem. Basically every crate that depends on `image` would add this extra profile (probably with different names too!), and crates that depend on those crates would also add them, and so on. Going forward with this RFC (without the other one first) will likely lead to lots of people sticking extra duplicated overrides in their `Cargo.toml`, which bothers me. An "upstream-down" approach strikes me as much nicer for the ecosystem as a whole.

On a somewhat separate tangent: how often do we expect users to actually add separate profiles, rather than just adding overrides for the `debug` profile? I personally think (though I may be wrong) that the primary use-case for this is compiling specific dependencies with opts in debug mode, in which case it's unclear we even need custom profiles, and not just "being able to specify overrides for existing profiles". Do we have a compelling use-case for where an entirely new profile is warranted?

Manishearth commented:
@jonhoo FWIW this RFC does propose allowing profiles in .cargo/config, but it also allows them in Cargo.toml for consistency

matklad commented:
@jonhoo note that the RFC supports overrides for build-in profiles, so I don't believe that "every crate that depends on image would add this extra profile". Rather, every crate which depends on `image` would add `[profiles.dev.overrides.image] opt-level = 3`. This indeed will be duplicated, but the amount of duplication is small, and gives you extra flexibility as each crate can decide precisely on the settings for image.

In general, I think the need to customize profiles should be rare, so I'd go for the increased flexibility rather than for increased convenience (but the syntax proposed in this RFC looks pretty and convenient). 


>I personally think (though I may be wrong) that the primary use-case for this is compiling specific dependencies with opts in debug mode, in which case it's unclear we even need custom profiles, and not just "being able to specify overrides for existing profiles". 

Hm, interesting! I believe that it is indeed possible to split this RFC further into "overrides" and "custom profiles" parts, and I agree that "overrides" part has potentially many more users, because fully custom profiles a required mainly by large users like Servo. My personal preference would be to nevertheless pursue the custom profiles part. I really like how it creates a very simple mental model for profiles: "if you pass `--profile foo`, then everything uses the `foo` profile". This model is pretty different for the one for build-in profiles, where different parts of the packages use different profiles. And having such simple model implemented in Cargo would be very useful, because it would make "fixing" build-in profiles to use this model simpler.

matklad commented:
@rfcbot fcp merge

matklad commented:
@rust-lang/cargo discussed this RFC at today's meeting, and it looks really great!

There are couple of implementation-level concerns/questions though.

matklad commented:
@rfcbot concern workspaces 

Cargo supports building several packages together as a part of the workspace. In workspaces, only the profile section from the root of the workspace (virtual or not) is taking an effect. Currently this works as expected, because profiles apply to all crates in the DAG of dependencies equally.

However with this RFC there's an observable difference between the "current" crate and all other crates. How this interacts with multi-package workspaces.

For example, if workspace consists of `foo` and `bar`, and `foo` depends on `bar`, and profiles section contains

```
[profile.dev]
opt-level = 0

[profile.dev.overrides."*"]
opt-level = 2
```

what rule applies to `bar`?

matklad commented:
@rfcbot concern build_override

`build_override` definitely feels like a special case. Would it be feasible/worthwhile to make a general feature which allows to tweak each `target`? The staraman syntax would be `[profile.dev.overrides.foo.build]`, `[profile.dev.overrides.foo.bin."*"]`, `[profile.dev.overrides.foo.test.test_bar]`?


I personally also think that maybe we should just always compile `build.rs` files with optimizations? You compile build.rs only once, so it seems like a win, and is somewhat similar to `cargo install` in spirit. There's also a question of what to do with crates like LALRPOP: if you use it, you definitely want to compile build.rs in release mode, and, with current proposal, every reverse dependecy of LALRPOP would need to specify this explicitly. 

It would be great if someone could measure the impact of always-release `build.rs`. How frequent are the cases when it significantly worsens compile times? How frequent are the cases where compile times are improved?

matklad commented:
Implementation-wise, it would probably make sense to file all of

- custom profile
- profile overrides
- profiles in .cargo/config

under separate feature-flags. 

We can also stabilize this piece-wise. For example, `overrides."image=0.2 piston>5.0"` looks more controversial than `overrides.image`, so probably could stabilize the second form first, and than make a follow-up.

joshtriplett commented:
@rfcbot concern Document how to set panic

This isn't a showstopper, just something I'd like to see documented in the RFC.

You *can't* currently pass custom options to rustc, so this doesn't apply to anything else. But right now, you might write a crate that *only* makes sense with `panic=unwind`, or *only* makes sense with `panic=abort`. You might set those in your build profiles. If someone overrides those profiles, they'll have problems.

I've seen multiple proposals for other ways to specify panic behavior, such as module attributes within Rust code. I don't know the current status of those proposals.

I'd like to ask that the RFC suggest how to handle this situation. Or, if the current state is that there's no other stable way to specify this, I'd like to suggest that an RFC get kicked off to provide a mechanism for this. (That RFC would not need to complete before this one gets accepted, but I'd like to see something proposed.)

The analogous situation for C: running `make CFLAGS='-O0 -g'` should override the Makefile specifying `-O3`, but should not override the Makefile specifying `-Wall` or `-fno-strict-aliasing` or `-std=c11`. People have rough conventions and understandings for which flags a command-line-specified `CFLAGS` should override and which it shouldn't (though this still remains the realm of undocumented tribal knowledge). Similarly, overriding a build profile should not cause a crate to fail to build.

joshtriplett commented on text/0000-custom-cargo-profiles.md:
```diff
@@ -0,0 +1,141 @@
+Custom profiles _can_ be listed in a `.cargo/config`, however the user is responsible for
+clearing up build directories if the profile changes. That is, it is undefined behavior
+to run `cargo build --profile foo` if `foo` has been defined in `.cargo/config` and the
+profile has been edited since the last time you ran `cargo build --profile foo`.
```
How hard would it be for us to detect profile changes and automatically rebuild in that case, just as we would if you changed `Cargo.toml`?

If it's easy, can we just do it?

If it's hard, can we list it as an alternative in the RFC and explain why we're not doing it?

Manishearth replied:
It's not exactly hard, just tricky I think. We could probably do it, but I thought as a first pass it's better to leave it undefined and in the future we can experiment with a scheme for this. This introduces a dependency to files out of the project root which is kinda icky.

joshtriplett replied:
We don't have to introduce a dependency on the *file*. What if we just remember the profile information used for the last compile, and if that changes, rebuild?

Manishearth replied:
Works. I can fix that.

alexcrichton commented:
The RFC currently mentions:

> In case of overlapping rules, the last mentioned rule will be applied.

How does this work with `.cargo/config`, though? What's the precedence of `.cargo/config` sections vs `Cargo.toml`? (and `.cargo/config` sort of has its own set of precedences as well...)

Manishearth commented:
> I personally also think that maybe we should just always compile build.rs files with optimizations? 

This is questionable. From-scratch build times _are_ important, and in Firefox's case we spend quite a bit of time build-time compiling the AST crates (syntex/aster/syn/whatever, I'm not sure which one is involved) in release mode. The ideal setup for Firefox is to compile bindgen in release mode, but compile some of its dependencies in debug mode (Rust codegen is not the bottleneck for bindgen). Or something. Being able to customize this would be great.

Also, in the general use case, `--debug` has multiple meanings: "I want to debug this" and "I want fast compile times". When using it in the latter case, there's no guarantee that it's the right thing to optimize build.rs. E.g. if build.rs has a dependency on the regex crate, it will spend forever compiling that deptree, even if you use regexes only once or twice. Small build.rs files are common.

I think we should expose some ability to customize build time profiles, even if we're not exposing everything.

matklad commented:
>This is questionable. From-scratch build times are important, and in Firefox's case we spend quite a bit of time build-time compiling the AST crates (syntex/aster/syn/whatever, I'm not sure which one is involved) in release mode. The ideal setup for Firefox is to compile bindgen in release mode, but compile some of its dependencies in debug mode (Rust codegen is not the bottleneck for bindgen).

But which default is better? I.e., if we flip build.rs default to `--release`, would the from-scratch build of firefox be slower, because we have compile AST crates in release, or faster, because bindgen is compiled in release as well?

Manishearth commented:
@joshtriplett @matklad 

> rfcbot concern workspaces

I have no opinion on how this should be resolved. I think applying it to the current crate as if there was only one crate in the workspace is fine. We could expose a way to customize this on a per-current-crate basis, idk.

> rfcbot concern build_override

The strawman seems to introduce a phantom "build" profile? I have no strong opinions here -- if the team prefers it that way I'm happy to merge in those changes.

> Implementation-wise, it would probably make sense to file all of ... under separate feature flags

👍 

> rfcbot concern Document how to set panic

IMO crates should be allowed to say `panic-strategy=strategy` where `strategy` is `abort`, `unwind`, or `all` (default) in their Cargo.toml. Using profiles as a two way thing (both to set build settings for a toplevel crate and constrain toplevel build settings from a dependency) seems like an overloading of the feature. This should be a simple RFC which we could mentor someone new on working.

Manishearth commented:
> How does this work with .cargo/config, though? What's the precedence of .cargo/config sections vs Cargo.toml? (and .cargo/config sort of has its own set of precedences as well...)

Probably Cargo.toml overriding .config, and .config following its own precedence.

The use case for firefox and custom build systems is that we can now autogenerate the profiles and Cargo.toml no longer needs profiles. In this case it doesn't matter.

The use case for the general user is so that they can globally tweak their debug/release profile (perhaps to set codegen-units globally, or some such), or introduce a "foobar" profile that does special things. In the former, it's not a big deal if Cargo.toml overrides it, but if we flipped the precedence it would be problematic: If Cargo.toml sets things like panic=abort that the crate _must_ use and that gets overridden, you have to add a _second_ .config file locally that copies the profile from Cargo.toml, and keep that in sync. Ick. In the latter use case the precedence doesn't matter at all, unless there's a clash (which I consider an edge case)

Manishearth commented:
@matklad 

> But which default is better? I.e., if we flip build.rs default to --release, would the from-scratch build of firefox be slower, because we have compile AST crates in release, or faster, because bindgen is compiled in release as well?

Overall faster. It can be made _even faster_ by cherry picking the crates (namely, compiling big AST crates in debug mode).

I think build_override being release is a sensible _default_, but I'd still like the ability to configure that.

Manishearth commented:
Pushed updates addressing the `.cargo/config` and workspaces issues.

@joshtriplett So I thought more about the `panic=abort` problem, and really I think it's very orthogonal to this RFC. You say:

> You might set those in your build profiles. If someone overrides those profiles, they'll have problems.

But profiles for dependencies are ignored anyway; so those profiles are already overridden.

I don't see how this RFC affects the status quo there. 

I'm more than happy to specify a rough proposal for this, I'm just not sure _where_ to fit it in the RFC (under "unresolved questions"?). If you have an idea let me know.

Manishearth commented:
So @matklad and I had a discussion where we realized that having `build_override` affect dependencies that are both regular and build dependencies may be surprising since it will affect the final binary. I changed the conflict resolution to use a priority order (specific rules win over `"*"`, and both of them win over `build_override`) over source order which fixes this.

Essentially, this means that:

 - `build_override` is "all crates which are _only_ transitive build dependencies (not regular transitive dependencies) and additionally do not have a specific rule"
 - `"*"` is "all crates that are not this crate (or workspace members) which do not have a specific rule"

matklad commented:
@rfcbot resolved build_override

An interesting peculiarity about `build_override` is that it affects not only the build script itself, but also its dependencies. So even if we invent some target-specific scheme, it won't help us with the build scripts. So it makes sense to add a special case for them!

Manishearth commented:
To expand on @matklad's comment, we realized that the "target" stuff was in the sense of the build.rs "crate" file, specifically, which is a "target" much like main.rs, examples, and other binaries. This is not the same as build_override. Providing the ability to customize single targets and/or build script binaries (as in, build.rs itself and not its dependencies) is interesting but it's not clear if it's necessary. It can be added backwards compatibly later, but I'm receptive to adding it now if folks think there's a need.

matklad commented:
Another panic releated question: what happens in this case

```
[profiles.dev.foo]
panic = "unwind"

[profiles.dev.bar]
panic = "abort"
```

That is, what should happen if we try to link two crates with different panic strategies?

Manishearth commented:
We can either let it explode at build time (which is fine imo) or we can force the `panic` key to be only toplevel.

(I prefer the latter, will fix that in a commit)

joshtriplett commented:
@Manishearth 
> But profiles for dependencies are ignored anyway; so those profiles are already overridden.

I was unaware of that. Seems like we should do something about that somehow. But given that, I think it'll suffice to just mention it as an unresolved issue.

Manishearth commented:
Done!

I think it's a simple, likely uncontroversial RFC, I could open it but it seems like an opportunity to mentor someone else to do so, might do that.

Manishearth commented:
rfcbot seems to have not understood the resolved command.

(also, the workspace one should be resolved as well i think)

joshtriplett commented:
@rfcbot resolved Document how to set panic

https://github.com/rust-lang/rfcs/pull/2282#issuecomment-362652695

joshtriplett commented on text/0000-custom-cargo-profiles.md:
```diff
@@ -0,0 +1,157 @@
+- Bikeshedding the naming of the keys
+- The priority order when doing resolution
+- The current proposal provides a way to say "special-case all build dependencies, even if they are regular dependencies as well",
+  but not "special-case all build-only dependencies" (which can be solved with a `!build_override` thing, but that's weird and unweildy)
```
s/unweildy/unwieldy/

matklad commented:
@rfcbot resolved workspaces

Treating `"*"` as `non-members` seems fine. I think the current syntax works fine both for workspaces with a virtual root and with a package root. 

However, is `"*"` the best symbol for dependencies? I think one may naturally assume that `*` means everything. Should we change it to `[profile.dev.overrides.":deps"]` or something similarly more explicit?

aturon commented:
@Manishearth just wanted to let you know that I had a good chat with @wycats today about where we want to head with profiles overall, which probably *will* want to have some impact on this RFC. I'll be writing up those thoughts ASAP.

matklad commented:
@aturon, @wycats ping :-) We need your thoughts here, because so far the RFC was talking primarily about the narrow "compiler options" aspect of profiles, and the wider "workflows" aspect is currently missing from the discussion.

aturon commented:
Thanks @matklad; I've been super tied up with futures work but will try to write some general thoughts here tomorrow.

aturon commented:
So I don't have time to write up my full thoughts on profiles in general -- I plan to write some blog posts about that in the coming weeks. But I'll try to summarize the bits that are most relevant to this RFC.

Today's profile settings basically boil down to three things: optimization level, debuginfo, and panic strategy. The settings of each of these knobs varies *by workflow* (though the current mapping is broken in some ways). And only the settings of the root crate matter.

I'm working on a re-conceptualization of this setup, both to clarify it and allow for better extensibility:

- Deprecate "profiles" as they are today.
- Instead, make "workflows" into a more first-class concept.
  - Like git's porcelain vs plumbing, rationalize Cargo as a set of primitive operations and a set of workflows on top. (This is also important for build system integration work).
  - Cargo comes with several built-in workflows (`check`, `test`, `build`, `run`)
  - We provide a more structured way to add workflows (like subcommands today, but further integrated)
- Today's "profiles" are instead "workflow configuration"
  - This greatly clarifies the fact that only the root crate's configuration matters; the *workflow* comes from the root, and the workflows defined in some library crate are only relevant for *developing* it.
- Workflows probably want some inter-relationship; we need to account for things like `--release` as well.

Now obviously the above is pretty hazy, but hopefully it gives you some idea of where at least @wycats and I would like to head.

Given that context, I don't feel great about adding custom profiles; I am really hoping we address that through workflows instead.

However, the more pressing issue tackled by this RFC, of customizing upstream profiles, seems orthogonal and great to explore regardless; we'd want it in any workflow customization anyway.

So I'm wondering if we can pare the RFC down to just controlling upstream profiles? And there, I wonder if we can *start* with just named crates, rather than full-on semver ranges?

Manishearth commented:
> Given that context, I don't feel great about adding custom profiles; I am really hoping we address that through workflows instead.

Can we continue to allow profiles to be overridden by .cargo/config? Because for Firefox's use case we need either .cargo/config to work, or custom profiles (ideally both).

Removing the semver ranges sounds ok.

Manishearth commented:
Updated. I removed the notion of custom profiles, however I kept around the ability to define profiles in `.cargo/config`. That can be made into a separate RFC if desired, it seems to be a small feature.

aturon commented:
@Manishearth Ah sorry, yes the `.cargo/config` seems fine.

aturon commented:
Thanks @Manishearth, looks great! I've checked off my review box; just waiting on @joshtriplett and @withoutboats now.

Manishearth commented:
uh oh it's the robot rebellion

matklad commented on text/0000-profile-dependencies.md:
```diff
@@ -0,0 +1,147 @@
+# Summary
+[summary]: #summary
+
+Allow overriding profile keys for certain dependencies, as well as providing
```
This sentence is :)

matklad commented:
We've discussed the broader issue of workflows at today's meeting, here's the summary. 

Today's profiles represent two different things. The first one is the workflow, or operation, which Cargo currently is doing, like testing, documenting or benchmarking. The other, more narrow, is the set of "optimization flags" applied when building the code.

It makes sense to split the current profile concept into two separate ones, by introducing **workflow** as a first-class thing for describing a particular set of operations cargo does, and scaling down **profile** concept to just mean the set of optimization flags. 

One can imagine that there will be roughly one workflow for each major Cargo command (testing workflow, documenting workflow, benchmarking workflow). Ideally, it should be possible to add custom workflows like a fuzzing workflow.

In contrast, there should be relatively fewer profiles: `dev` and `release` seem to cover most use-case, but obvious possible extensions include `release-with-debuginfo` and a "slightly optimized" dev. 

Each workflow should be linked to some default profile, but the profile could be overridden via command line flag. For example, testing workflow would use `dev` profile by default, and benchmarking workflow would use the `release` profile, however, `test --profile=release` would also work.


Now, the interesting question is, which flags belong to profiles, and which flags belong to workflow? The interesting dimension for comparison here is "can I apply this flag to only part of the packages?". It is interesting because workflows are global for the whole project, while profiles, with this RFC, can apply only to the specific packages. However, certain optimization flags, like `lto` apply to the artifact as a whole: i.e, their granularity is larger than a crate, but smaller than project. So, this gives us three bins to sort flags into:

* workflow flags (per project)
* profile global (per artifact)
* profile local (per package == per dependency == per library crate)

Sorting the current flags gives this picture:

* workflow:
  - rpath
  - panic

* profile global:
  - lto
  - codegen-units

* profile local:
  - opt-level
  - debug
  - debug-assertions

Manishearth commented:
Why is codegen-units profile global?

I think LTO, rpath, panic are all global, of which rpath/LTO are workflow-related ones, and panic is weird.

We can amend this RFC so that LTO, rpath, panic can only be specified at the top level

Manishearth commented:
@matklad I'm confused as to what the status of this RFC is. Was your comment an indicator of the overall plans for profiles, or a request for amendments to this one (seems like they would be minor). The FCP has finished already.

matklad commented:
@Manishearth this was a comment about the general future direction of profiles, the RFC is great as it stands! I think it should be merged now, but I am not sure about the precise process to do this :)

Manishearth commented:
Cool. I think we're just supposed to rename the file and fill in details and then merge it. I'll let @nrc merge

Manishearth commented:
Updated

Manishearth commented:
https://botbot.me/mozilla/rust-internals/2018-03-02/?msg=97436474&page=5
